### ISSUE TITLE
Ensure cell clustering process can be run with a variety of cluster types

### ISSUE THREAD

**Issue by alex-l-kong** [angelolab/ark-analysis#881]
**For future reference**

Use draw.io for flowcharts lol.

**Relevant background**

The cell clustering pipeline currently only supports pixel clusters generated from `2_Pixie_Cluster_Pixels` (`pixel_som_cluster` or `pixel_meta_cluster_rename`). This has worked fine because `3_Pixie_Cluster_Cells` is designed to run immediately afterwards.

However, increasingly users have needed to run cell clustering on inputs outside of pixel clustering. The cell clustering pipeline will need to be generalized in order to support this.

**Design overview**

Generalizing cell clustering will require changes to multiple components.

We're currently presetting the following variables using `cell_clustering_params` saved from `2_Pixie_Cluster_Pixels`:

- fovs
- channels
- segmentation_dir
- seg_suffix
- pixel_data_dir
- pc_chan_avg_som_cluster_name
- pc_chan_avg_meta_cluster_name

We should still provide support for this, however we should also give the user flexibility to use custom inputs.

_Pixel SOM input data (pixel_data_dir)_

Currently, the input data to train the SOM is generated using a helper function. However, the user will need to be specifying this data on their own.

The pixel clustering SOM input data is formatted as such:

| cell_size | fov | segmentation_label | pixel_som_cluster_0 | pixel_som_cluster_1 | ... |
| ---------| --- | -------------------- | --------------------- | ------------------- | --- |
| ... | fov0 | 0 | ... | ... | ... |
| ... | fov0 | 1 | ... | ... | ... |
| ... | ... | ... | ... | ... | ... |

For the purposes of generalizability, we can replace columns `pixel_som_cluster_0`, `pixel_som_cluster_1`, etc. with `user_defined_col_0`, `user_defined_col_1`, etc. Users will need to provide data that contains, at the very least, the `fov`, `segmentation_label`, and these user defined column data.

In the case of custom inputs for cell clustering, the user will need to specify these column names on their own, as they will not have consistent a prefix naming scheme.

This change will mean that if the full Pixie pipeline is run, the cell SOM inputs should be created as an explicit preprocessing step to training the actual cell SOM (as opposed to being implicitly built into the training function). This will allow the training function to use a consistent form of input data for both the full Pixie and the generalized pipeline.

_Cluster channel average expression file_

If the user is providing generalized inputs, they should not be required to provide this file, as this is based on pixel clusters only and has no reasonable interpretation outside of this context.

_Weighted cell channel average file_

If the user is running this as part of the full Pixie pipeline, this file should also be generated as an explicit preprocessing step before training the cell SOM (and not inherently built into that function, as it is now).

Users running this with generalized inputs will not need to generate this file.

_Deprecating cell clustering params (for those starting with cell clustering)_

Users starting directly from cell clustering should not be required to make their own cell clustering params JSON file.

Users should instead provide the path to their custom data, formatted like the table above. From this file, we can infer their list of FOVs and the columns to subset on (throwing an error if the data doesn't follow a consistent column naming scheme). Users will still need to provide a segmentation directory, segmentation suffix, and cell table path on their own.

**Code mockup**

For this process, we should make a new notebook called `3b_Generic_Cluster_Cells`. This is a tentative name subject to change.

_Data preprocessing (`3a_Pixie_Cluster_Cells`, tentative name)_

Users running the full Pixie pipeline process can use the outputs directly from `2_Pixie_Cluster_Pixels`, however the generation of inputs into the cell SOM pipeline will be different.

To support both pipelines (full Pixie and starting from cell clustering), `crete_c2pc_cell_data` should directly take the input data as opposed to generating it itself. In `3a_Pixie_Cluster_Cells`, this can happen with:

```
# user will set variables cell_table_path, pixel_cluster_col, and cell_som_input_name prior

cluster_counts, cluster_counts_size_norm = create_c2pc_data(
    fovs, pixel_data_dir, cell_table_path, pixel_cluster_col
)

# define the count columns found in cluster_counts_norm
cell_cluster_columns = cluster_counts_size_norm.filter(
    regex=f'{pixel_cluster_col}.*'
).columns.values

feather.write_dataframe(
    cluster_counts_size_norm,
    os.path.join(base_dir, cell_som_input_name),
    compression='uncompressed'
)
```

The variable `cell_cluster_columns` is needed to support a generic list of columns a user should now be able to specify to `train_cell_som` identifying the features needed for training (required for generalized cell SOM inputs). This will no longer be consistent with the `pixle_cluster_col_prefix` used in the past (which ran a regex find of all matching columns), so these arguments will need to be removed and replaced with the explicit set of columns passed.

Additionally, the weighted cell channel average file should also be precomputed. This can be done by using the unnormalized `cluster_counts`:

```
# user will be setting variables pixel_cluster_col and weighted_cell_channel_name prior,

# depending on which pixel_cluster_col is selected, choose the pixel channel average table accordingly
if pixel_cluster_col == 'pixel_som_cluster':
    pc_chan_avg_name = pc_chan_avg_som_cluster_name
elif pixel_cluster_col == 'pixel_meta_cluster_rename':
    pc_chan_avg_name = pc_chan_avg_meta_cluster_name

weighted_cell_channel = compute_p2c_weighted_channel_avg(
   pixel_cluster_channel_avg, channels, cluster_counts,
   fovs=fovs, pixel_cluster_col=pixel_cluster_col
)

feather.write_dataframe(
    weighted_cell_channel,
    os.path.join(base_dir, weighted_cell_channel_name),
    compression='uncompressed'
)
```

Users can then pass `weighted_cell_channel_name` as normal.

_Parameter setting (`3b_Generic_Cluster_Cells` tentative name)_

The new notebook for the generic cell clustering process will be structured similarly, however, one major difference will be in the way the inputs are set.

Users will need to provide an explicit path to their input data:

```
cell_som_input_data = '/path/to/input/data'
```

After this is loaded in, we can then determine the list of FOVs using:

```
cell_som_input_data['fov'].unique()
```

Users will need to provide their own list of columns to use:

```
cell_som_cluster_columns = [...]
```

They will also need to specify their own `segmentation_dir`, `seg_suffix`, and `cell_table_path`.

_Changes to the cell clustering process_

- `train_cell_som`: should now take `cell_som_input_name` as an argument to specify the input file. It should no longer take `pixel_data_dir`, `pc_chan-avg_name`, `cluster_counts_name`, `cluster_counts_size_norm_name`, or `pixel_cluster_col` as explicit arguments. `channels` should be renamed to `cell_som_cluster_columns`.
- Any function that takes `pixel_cluster_col` or `pixel_cluster_col_prefix` arguments to determine columns to subset on should now just take an explicit list of `cell_som_cluster_columns`.
- Any function that takes `weighted_cell_channel_name` should support a `None` value for this param. Generic cell cluster inputs will not have a good interpretation for average weighted channel expression which is specific to pixel clusters. When these functions are called from notebook `3b`, they should set this param to `None`, causing the average channel expression computations to be skipped
- `3b_Generic_Cluster_Cells`: the weighted cell channel heatmaps should not be generated per the previous point.

_Example dataset changes_

We'll need to add sample input data for notebook `3b`. This should consist of a custom `input_data` file, a `cell_table`, and `segmentation_data`.

_Notebook test changes_

For `3a`, ensure that the preprocessing cells get run. Changes should not need to be significant. The test class for this notebook should be renamed accordingly.

For `3b`, ensure that each cell runs with the example dataset for this notebook. We'll need a new test class for this.

_Updated function testing_

Most significant change here will be ensuring we define the full set of columns to pass in place of `pixel_cluster_col_prefix`. Additionally, conditionals where `weighted_cell_channel_avg` needs to handle the `None` case will also need additional testing.

**Required inputs**

_3a_Pixie_Cluster_Cells_

The existing inputs should integrate smoothly.

_3b_Generic_Cluster_Cells_

The user will need to explicitly specify their cell SOM input data, cell table, segmentation dir, and segmentation suffix.

**Output files**

_2_Pixie_Cluster_Pixels_

This will now generate the cell SOM input data and weighted cell channel average file on top of the cell clustering parameters file.

_3a_Pixie_Cluster_Cells_

Same as before, except where they will be generated may be different.

_3b_Generic_Cluster_Cells_

Mostly the same, with the exception of heatmaps, which will not be displayed.

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [ ] A couple days
- [ ] A week
- [X] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review:

January 31

(I'll can have the code without tests done earlier if Lorenz needs it before then)

Estimated date when the finalized project will be merged in: 

Early February

**Comment by alex-l-kong** [comment]
@ngreenwald @JLrumberger let me know if this looks good!

**Comment by ngreenwald** [comment]
@cliu72, the goal here is to use the Pixie pyflowSOM logic to cluster cells, and then the GUI to reassign cells, for multiple input types, not just pixel clustering. The immediate use-case is Lorenz, but in general I think it would be useful to have this ability. There are a couple options:

1. We make these changes to `ark`, so that the Pixie notebook 3 for cell clustering is set up to handle multiple input types, with prompts guiding the user for which ones are relevant
2. We make these changes, but not until after pinning a version for your paper that doesn't include them. For example, if you think this would distract the reviewers, or potential users who click on the code link from your paper, we can make a version that keeps things as is and doesn't reference other input data types. 
3. We make these changes under the hood, but we never update the Pixie cell clustering notebook in `ark` to reflect them. Instead, we have an entirely separate notebook for generic cell clustering that walks through the pipeline, and the Pixie notebook 3 remains the direct follow-up to notebook 2. 

As I typed these options out, I'm a bit leaning towards option 3, where we make like a notebook 3b, call it Pixie Cell cluster only (or something that rolls off the tongue better), to keep the logic isolated. But I can imagine any of these options working, it's up to you

**Comment by cliu72** [comment]
I agree with you @ngreenwald, I favor option 3. I think option 1 (basically what Alex has outlined here) would confuse people who read the paper and then go to try it out (basically what you wrote in 2 above). I think it would be confusing that some inputs (ex. pixel cluster average file) and outputs (ex. weighted expression heatmap) would no longer be relevant, depending on the input. I think it's better to contain pixel-based cell clustering in its own notebook (at least for now, while the paper is getting out there). I think 2 is an okay option too, but relies on the user to make sure they're referencing the write pinned version of the code. I think 3 is the cleanest/least confusing.

@alex-l-kong Regarding this part "Because we can no longer assume a constant naming scheme, we should instead infer these column names by dropping the first n values or so. How many columns will be attached at the beginning will be a topic for further discussion." - I think you can just require the user to specify the column names they want (kind of like how we ask the user to specify the channels they want in the pixel clustering notebook). It seems cleaner than trying to drop some arbitrary number of columns.

**Comment by ngreenwald** [comment]
Okay, sounds good. Alex, can you update the design doc to reflect this? We'll still move the pixel clustering aggregation step out of the cell_clustering function and into the pixel notebook, but we won't need logic throughout the cell clustering notebook that deals with reading input data of different types, controls what plot types, etc. 

It may be the case that you want to change the API for the cell clustering function to make it easier to handle different input types, but within each notebook, the argument should be pre-populated so that the user doesn't need to deal with it.

**Comment by cliu72** [comment]
Why is the pixel clustering aggregation step going to be moved into the pixel notebook? I thought one of the advantages of separating them is that you don't need segmentation masks or a cell table to run the pixel clustering notebook - and that the pixel and cell parts are separate.

**Comment by ngreenwald** [comment]
You're right, we need to remove it from the main `train_som` function, but it actually makes more sense keep it in the cell clustering notebook, rather than the end of the pixel clustering notebook.

**Comment by alex-l-kong** [comment]
@ngreenwald @cliu72 ok we'll go with option 3, that's actually a huge relief 😅. Definitely has a much more intuitive workflow.

**Comment by ngreenwald** [comment]
Agreed. Tag us once you've updated so we can confirm, then should be good to go

**Comment by alex-l-kong** [comment]
@ngreenwald @cliu72 @JLrumberger updated the design doc to reflect a new notebook for generic cell clustering. Also moves preprocessing of Pixie cell SOM inputs and weighted cell channel file to the Pixie cell clustering notebook.

**Comment by ngreenwald** [comment]
> Any function that takes weighted_cell_channel_name should support a None value for this param. Generic cell cluster inputs will not have a good interpretation for average weighted channel expression which is specific to pixel clusters. When these functions are called from notebook 3b, they should set this param to None, causing the average channel expression computations to be skipped

Which functions are these? If they don't need to be run at all for 3b, can we just remove them? Or are there examples that perform multiple tasks, only one of which depends on the weighted channels?

In order to break this into smaller pieces, I think the first PR should just tackle the changes to the existing notebook 3, pulling out the computation from the main training loop into a preprocessing helper. Then, branching off that PR, you can open subsequent ones to address the additional notebook.

**Comment by alex-l-kong** [comment]
@ngreenwald they're functions that perform multiple tasks, one of which depends on the weighted channels.

Agreed that the existing notebook 3 changes should be handled first. We can address how to handle the weighted channels computations in the PR that adds the new notebook.

**Comment by cliu72** [comment]
This looks good to me

**Comment by alex-l-kong** [comment]
#892 has addressed this. However, #850 may still have have pending issues (ex. turning off normalization for overall abstraction process).

Can we close this one and leave #850 open?

**Comment by ngreenwald** [comment]
sounds good
